### PR TITLE
API-1097: Fetch errors from ES index for the GetConnectionBusinessErrors endpoint

### PR DESCRIPTION
**Pull request by pierrejolly**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public function test_it_returns_an_empty_array_when_nothing_is_indexed(): void
+    {
+        $result = $this->selectLastConnectionBusinessErrorsQuery->execute('erp', '2020-01-07', 10);
+        Assert::assertEquals([], $result);
+    }
+
+    public function test_it_returns_the_last_business_errors_of_a_connection(): void
+    {
+        $this->esClient->bulkIndexes([
+            // Ignored: error is too older (more than 7 days)
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php:31]:
```suggestion
            // Ignored: error is too old (more than 7 days)
```

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            [
+                'connection_code' => 'erp',
+                'error_datetime' => '2019-12-31T00:00:00+00:00',
+                'content' => ['message' => 'Error 1'],
+            ],
+            // Ignored: 3rd result (oldest) on a $limit if 2
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php:37]:
```suggestion
            // Ignored: 3rd result (oldest) on a $limit of 2
```

File: src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                'connection_code' => 'erp',
+                'error_datetime' => '2020-01-06T00:00:00+00:00',
+                'content' => ['message' => 'Error 4'],
+            ],
+            [
+                'connection_code' => 'erp',
+                'error_datetime' => '2020-01-07T00:00:00+00:00',
+                'content' => ['message' => 'Error 5'],
+            ],
+            // Ignored: error is younger than the $endDate param
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/tests/Integration/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQueryIntegration.php:59]:
```suggestion
            // Ignored: error is newer than the $endDate param
```

File: src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php [hunk 4]
```diff
@@ -96,30 +96,30 @@ public function index($id, array $body, Refresh $refresh = null)
...
-     * @param string       $keyAsId
+     * @param ?string      $keyAsId
      * @param Refresh|null $refresh
      *
      * @throws MissingIdentifierException
      * @throws IndexationException
      *
      * @return array see {@link https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/_indexing_documents.html#_bulk_indexing}
      */
-    public function bulkIndexes($documents, $keyAsId, Refresh $refresh = null)
```
Comment (pchasle) [inline src/Akeneo/Tool/Bundle/ElasticsearchBundle/Client.php:107]: Can you remove this file from the PR?

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        $this->esClient = $esClient;
+    }
+
+    public function execute(string $connectionCode, string $endDate = null, int $limit = 100): array
+    {
+        [$from, $to] = $this->getDateTimeInterval($endDate);
+
+        $result = $this->esClient->search([
+            'query' => [
+                'bool' => [
```
Comment (pchasle) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php:32]: Could you add the `constant_score` property has suggested by Samir ? :)

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+use Akeneo\Connectivity\Connection\Domain\ErrorManagement\Model\Read\BusinessError;
+use Akeneo\Connectivity\Connection\Domain\ErrorManagement\Persistence\Query\SelectLastConnectionBusinessErrorsQuery;
+use Akeneo\Tool\Bundle\ElasticsearchBundle\Client;
+
+/**
+ * @author Pierre Jolly <pierre.jolly@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class ElasticsearchSelectLastConnectionBusinessErrorsQuery implements SelectLastConnectionBusinessErrorsQuery
```
Comment (willy-ahva) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php:16]: Just a suggestion, replace Elasticsearch by ES? We already do it with DBAL.

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                                ]
+                            ]
+                        ],
+                    ],
+                ],
+            ],
+            'sort' => [
+                'error_datetime' => 'desc',
+            ],
+            'size' => $limit,
```
[the remark this answers is not in the document]
Comment (pierrejolly) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php:53]: As seen if we use `terminate_after` with `sort` it's doesn't work as expected in this case

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        ]);
+
+        return array_map(function (array $row) {
+            $data = $row['_source'];
+
+            return new BusinessError(
+                $data['connection_code'],
+                \DateTimeImmutable::createFromFormat(\DateTimeInterface::ATOM, $data['error_datetime'], new \DateTimeZone('UTC')),
+                json_encode($data['content'])
```
[the remark this answers is not in the document]
Comment (pierrejolly) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php:62]: I don't understand, I need the `connection_code`, the `error_datetime` and the `content` (everything in fact)

[the remark this answers is not in the document]
Comment (willy-ahva) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Elasticsearch/Query/ElasticsearchSelectLastConnectionBusinessErrorsQuery.php:68]: yes my bad

**Pull request merged**